src/transforms/key_value_parser.rs

```diff
@@ -0,0 +1,409 @@
+    pub separator: String,
+    pub target_field: Option<Atom>,
+    pub trim_key: Option<String>,
+    pub trim_value: Option<String>,
+    pub types: HashMap<Atom, String>,
+}
+
+inventory::submit! {
+    TransformDescription::new::<KeyValueConfig>("key_value_parser")
+}
+
+#[typetag::serde(name = "key_value_parser")]
+impl TransformConfig for KeyValueConfig {
+    fn build(&self, _cx: TransformContext) -> crate::Result<Box<dyn Transform>> {
+        let field = self
+            .field
+            .as_ref()
+            .unwrap_or(&event::log_schema().message_key());
+
+        let conversions = parse_conversion_map(&self.types)?;
+
+        let trim_key = if let Some(key) = &self.trim_key {
+            Some(key.chars().collect())
+        } else {
+            None
+        };
+
+        let trim_value = if let Some(val) = &self.trim_value {
+            Some(val.chars().collect())
+        } else {
+            None
+        };
+
+        Ok(Box::new(KeyValue {
+            conversions,
+            drop_field: self.drop_field,
+            field: field.clone(),
+            field_split: Option::from(self.field_split.clone()),
+            overwrite_target: self.overwrite_target,
+            separator: self.separator.clone(),
+            target_field: self.target_field.clone(),
+            trim_key,
+            trim_value,
+        }))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn output_type(&self) -> DataType {
+        DataType::Log
+    }
+
+    fn transform_type(&self) -> &'static str {
+        "key_value_parser"
+    }
+}
+
+pub struct KeyValue {
+    conversions: HashMap<Atom, Conversion>,
+    drop_field: bool,
+    field: Atom,
+    field_split: Option<String>,
+    overwrite_target: bool,
+    separator: String,
+    target_field: Option<Atom>,
+    trim_key: Option<Vec<char>>,
+    trim_value: Option<Vec<char>>,
+}
+
+impl KeyValue {
+    pub fn parse_pair(&self, pair: &str) -> Option<(Atom, String)> {
+        let pair = pair.trim();
+        let field_split = self
+            .field_split
+            .as_ref()
+            .ok_or_else(|| Some("".to_string()))
+            .unwrap();
+
+        let fields = if field_split.is_empty() {
+            let mut kv_pair = pair.split_whitespace();
+            let key = kv_pair.next()?;
+            let val = kv_pair.next()?;
+            if kv_pair.next().is_some() {
+                error!(
+                    message = "KV parser saw more than one separator",
+                    rate_limit_secs = 30
+                );
+                return None;
+            }
+
+            (key, val)
+        } else {
+            let split_index = pair.find(field_split).unwrap_or(0);
+            let (key, _val) = pair.split_at(split_index);
+            let key = key.trim();
+            let val = pair[split_index + field_split.len()..].trim();
+
+            if key.is_empty() || val.is_empty() {
+                return None;
+            }
```

Comment:
Is `"key="` (empty value) valid here?

Reply:
Currently, that key would be dropped, but the transform would succeed with all other key/values. [There's a test for it](https://github.com/timberio/vector/pull/3284/files#diff-cb98402adc309a1927a810f3e4d88de7R334). 

So yes, the following message would be valid:
`name=Nick, age=, job=software`

Would result in:
```
{
  "name": "Nick",
  "job": "software"
}
```

This was intentional. I'm open to whatever you think is best but, I would prefer the whole transform not fail on this scenario where it's _mostly_ parse-able.

Reply:
Right, the question is if this should result in `"age": ""` or `"age": null`. I personally think that would be less surprising because it doesn't throw away incoming data (the key), but I'm not the target user for this transform.

Reply:
Ahh, that is a good point. I agree, but I'm not sure what to go with. I'm leaning `null` because an empty string implies type. I'll move in that direction but, open to opinions or changing it in the future.

Reply:
Empty string was easier. I don't know how much it's going to matter. I'm using Elasticsearch for a sink so, I'll see what happens there if it creates problems.

Reply:
I'd tend towards `null` as well, but if empty string is easier then go with that. We can figure out how best to change the behavior if it actually causes a problem with real data.